### PR TITLE
197: add oauth2 to poster logic

### PR DESCRIPTION
**Pull request by vladarefiev**
PR for https://github.com/timelesslounge/timelessis/issues/197
- added oauth2 to poster api logic, getting auth_token
- fix tests

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#249](https://github.com/timelesslounge/timelessis/issues/249) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#249](https://github.com/timelesslounge/timelessis/issues/249) is assigned to @vryazanov[/z](https://www.0crat.com/u/vryazanov), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/191220cd-1a1f-4eb4-a303-dce4086b9ee6); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: timeless/poster/api.py [hunk 1]
```diff
@@ -56,34 +57,42 @@ def send(self, **kwargs):
...
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
 
-    def access_token(self):
-        """Fetches authorization tokens
-        :return: Token retrieved from Poster api or error returned by poster pi
+class Authenticated:
+    """ Poster Auth class """
+    def __init__(self, auth_data: PosterAuthData, **kwargs):
+        self.auth_url = "https://joinposter.com/api/v2/auth/access_token"
```
Comment (vryazanov) [inline timeless/poster/api.py]: @vladarefiev let's move this to class attribute

File: timeless/poster/api.py [hunk 2]
```diff
@@ -56,34 +57,42 @@ def send(self, **kwargs):
...
-            method=self.POST,
-            action="auth/access_token"
-        )
+        Authentication into poster API
+        https://dev.joinposter.com/en/docs/api#authorization-in-api and use
+        """
+        auth_data = {
+            "application_id": self.auth_data.application_id,
+            "application_secret": self.auth_data.application_secret,
+            "grant_type": 'authorization_code',
```
Comment (vryazanov) [inline timeless/poster/api.py]: @vladarefiev  `"` instead of `'`, otherwise linter will fail

File: tests/integration/it_poster_test.py [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
+        code="test_code",
+    )
+    auth_token = "test_auth_token"
 
+    def auth(*args, **kwargs):
+        mocked = mock.Mock()
+        mocked.return_value = auth_token
+        return mocked
+
+    with mock.patch.object(Authenticated, 'auth', auth):
```
Comment (vryazanov) [inline tests/integration/it_poster_test.py]: @vladarefiev It thinks it's a bit artificial test. Let's mock response from server. WDYT?

File: timeless/poster/api.py [hunk 4]
```diff
@@ -56,34 +57,42 @@ def send(self, **kwargs):
...
+            "redirect_uri": self.auth_data.redirect_uri,
+            "code": self.auth_data.code,
+        }
+
+        response = requests.post(self.auth_url, data=auth_data)
+
         if not response.ok:
             raise Exception("Problem accessing poster api")
```
Comment (vryazanov) [inline timeless/poster/api.py]: @vladarefiev mb we will introduce a special expection type for errors related to Poster APIs? For example `class PosterAPIError(Exception)`

File: timeless/poster/api.py (hunk 1 above)
Comment (vladarefiev) [reply timeless/poster/api.py]: @vryazanov good catch, agree

File: tests/integration/it_poster_test.py (hunk 3 above)
Comment (vladarefiev) [reply tests/integration/it_poster_test.py]: @vryazanov Oops, you are stealing money from rultor :)

File: timeless/poster/api.py (hunk 4 above)
Comment (vladarefiev) [reply timeless/poster/api.py]: @vryazanov Hmm, it's not my stuff, but i totally agree with you.

**Comment by vladarefiev** [comment]
@vryazanov resolved comments. pls have a look

**Comment by vryazanov** [comment]
@vladarefiev you forgot to fix test :) actually there you test that mock works properly, not the logic inside `auth` function

**Comment by vladarefiev** [comment]
@vryazanov fixed

File: tests/integration/it_poster_test.py [hunk 5]
```diff
@@ -1,15 +1,30 @@
...
+        'Mock date',
+        (),
+        {
+            'ok': True,
+            'json': json_object,
+        }
+    )
+
+    auth_token = Authenticated(auth_data).auth()
+    assert auth_token == auth_token
```
Comment (vryazanov) [inline tests/integration/it_poster_test.py]: @vladarefiev what do you think about these:
```python
class Response:
    auth_token = "861052:02391570ff9af128e93c5a771055ba88"

    def ok(self):
        return True

    def json(self):
        return {"access_token": self.auth_token}

requests_mock.post.return_value = Response()
assert Authenticated(auth_data).auth() == Response.auth_token
```
it's clearer, I think.

Comment (vladarefiev) [reply tests/integration/it_poster_test.py]: @vryazanov like

**Comment by vladarefiev** [comment]
@vryazanov done, take a look pls

**Comment by vryazanov** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@vladarefiev a test is failing, please check

**Comment by vladarefiev** [comment]
@emilianodellacasa can't run this test locally, but now should be fixed

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by vladarefiev** [comment]
@0crat status

**Comment by 0crat** [comment]
Job was finished in 19 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by vladarefiev** [comment]
@ypshenychka for some reason 0crat didn't mention you.

**Comment by emilianodellacasa** [comment]
@vladarefiev @ypshenychka Bot can be slow, let's wait a couple of hours

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @vryazanov[/z](https://www.0crat.com/u/vryazanov), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#249](https://github.com/timelesslounge/timelessis/issues/249) is now out of scope

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/timelesslounge/timelessis/issues/249#issuecomment-463640095))

@vladarefiev This is what I know about this job in [CF9LG1YCA](https://www.0crat.com/p/CF9LG1YCA), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#249](https://github.com/timelesslounge/timelessis/issues/249) is not in scope
  * The job is waiting quality review verdict by @ypshenychka[/z](https://www.0crat.com/u/ypshenychka) for 1hr
  * @vryazanov[/z](https://www.0crat.com/u/vryazanov) will receive 20 minutes after quality review
  * Job [footprint](https://www.0crat.com/footprint/CF9LG1YCA?q=%7Bjob%3A%27gh%3Atimelesslounge%2Ftimelessis%23249%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +25 point(s) just awarded to @vryazanov[/z](https://www.0crat.com/u/vryazanov)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**